### DISCUSSION TITLE
Differences Homebrew on MacOS vs Linux

### DISCUSSION THREAD

**Discussion by oatbrew** [Homebrew/discussions#2631] [Linux] [answered]
To which extend are both different? They produce binaries for different operating systems, sure.
But are those build from the same sources? Does Brew on Linux use the same formulae as on MacOS?
In case it does would it be possible to fetch sources with Brew on Linux, transfer fetched packages to MacOS and build a MacOS binary?
Sources would be the same but build tools/compiler would be different?

**Comment by iMichka** [comment, marked as answer]
Hi. Yes the sources are exactly the same. All the packages build definitions are defined at https://github.com/Homebrew/homebrew-core.
You can look at each formula separately, some of them have OS.mac? / OS.linux conditions for minor mac/Linux only tweaks.

The big difference between macOS and Linux are mainly:
- mac bottles are built with system llvm/clang and linux bottles are built with system gcc 5 (we plan to migrate to gcc 7 next year).
- mac formulae try to use as many low-level system libraries as possible (gettext, curl, ncurses, ...), whereas Linux packages ship all of these (because docker and a more fragmented install base).

I do not think it is possible to cross-compile though: each platform has it's own build machine.

**Reply by oatbrew** [reply]
I do not mean cross compiling. 
Source packets e.g. would be downloaded on a Linux machine. Then transferred to a MacOS machine. MacOS tools (XCode) would then build the binary. Do you know if those  OS.Mac/OS.Linux conditions are usually related to *downloading* sources? In that case there may be the possibility that transferring sources to MacOS from Linux (or vice versa) could not work because maybe either of those expects a different packet version?

**Reply by SMillerDev** [reply]
No, the source is always the same between packages.

The only difference would be in software that has to bootstrap itself, example: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/go.rb

**Reply by gromgit** [reply]
There's also the issue of dependencies. I don't think there's a way to download macOS bottles on a Linux box, so unless I missed something, you're committing to building a lot more than just your chosen projects from source.
